core/src/Kokkos_BitManipulation.hpp

```diff
@@ -19,9 +19,36 @@
 
 #include <Kokkos_Macros.hpp>
 #include <Kokkos_NumericTraits.hpp>
+#include <climits>  // CHAR_BIT
+#include <type_traits>
 
 namespace Kokkos::Impl {
 
+template <class T>
+KOKKOS_FUNCTION constexpr T byteswap_fallback(T x) {
+  if constexpr (sizeof(T) == 1) {
+    return x;
+  } else {
+    using U = std::make_unsigned_t<T>;
+
+    size_t diff = CHAR_BIT * (sizeof(T) - 1);
+
+    U mask1 = static_cast<unsigned char>(~0);
+    U mask2 = mask1 << diff;
+    U val   = x;
+    for (size_t i = 0; i < sizeof(T) / 2; ++i) {
+      U byte1 = val & mask1;
+      U byte2 = val & mask2;
+
+      val = (val ^ byte1 ^ byte2 ^ (byte1 << diff) ^ (byte2 >> diff));
+      mask1 <<= CHAR_BIT;
+      mask2 >>= CHAR_BIT;
+      diff -= 2 * CHAR_BIT;
+    }
+    return val;
```

Comment:
```suggestion
  using U = std::make_unsigned_t<T>;
    U val = x; 
    std::size_t diff = CHAR_BIT * (sizeof(T) - 1);
    std::size_t byte_mask = static_cast<unsigned char>(~0);
    for (std::size_t i =0; i< sizeof(T)/2; ++i) {
      U low_mask  = byte_mask << i*CHAR_BIT;
      U high_mask = byte_mask << (sizeof(T)-1-i)*CHAR_BIT;
      U low_value = val & low_mask;
      U high_value = val & high_mask;
      val = (val & ~low_mask)  | (high_value >> diff);
      val = (val & ~high_mask) | (low_value << diff);
      diff -= 2* CHAR_BIT;
    }
    return val;
```
is clearer to me and seems to be better recognized by at least `clang`, see https://godbolt.org/z/zb4nEMPYv. I'm not quite sure how
```
val = (val ^ byte1 ^ byte2 ^ (byte1 << diff) ^ (byte2 >> diff));
```
works
```
val = (val & ~low_mask)  | (high_value >> diff);
```
clears the low bits first and then sets them to the desired value.

Reply:
I do understand how the @dalg24 version works.  While it may not look like it, this is just applying the regular xor trick for swapping two values.

`byte1 ^ (byte2 >> diff)` results in the low bits that need changing.

`val ^= byte1 ^ (byte2 >> diff)` changes the low bits.

The other two terms do the same thing for the high bits.

That being said, I also find the @masterleinad version easier to understand.

And if you are looking for more possible implementations, I believe this is the one from the paper proposing `byteswap`: https://github.com/bruxisma/byteswap/blob/master/include/ixm/byteswap.hpp

I'll approve it, but if the implementation changes, that's okay too.